verifier/verifier.go

```diff
@@ -0,0 +1,255 @@
+// Package Verifier provides an implementation of notation.Verifier interface
+package verifier
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/notaryproject/notation-go"
+	"github.com/notaryproject/notation-go/dir"
+	"github.com/notaryproject/notation-go/internal/envelope"
+	"github.com/notaryproject/notation-go/internal/plugin"
+	"github.com/notaryproject/notation-go/internal/plugin/manager"
+	"github.com/notaryproject/notation-go/internal/slice"
+	"github.com/notaryproject/notation-go/verifier/trustpolicy"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+)
+
+// verifier implements notation.Verifier
+type verifier struct {
+	trustPolicyDoc *trustpolicy.Document
+	pluginManager  pluginManager
+}
+
+// pluginManager is for mocking in unit tests
+type pluginManager interface {
+	Get(ctx context.Context, name string) (*manager.Plugin, error)
+	Runner(name string) (plugin.Runner, error)
+}
+
+// New a verifier based on local file system
+func New() (notation.Verifier, error) {
+	// load trust policy
+	policyDocument, err := trustpolicy.LoadPolicyDocument()
+	if err != nil {
+		return nil, err
+	}
+	return NewVerifier(policyDocument, manager.New(dir.PluginFS()))
+}
+
+// NewVerifier creates a new verifier given trustPolicy and pluginManager
+func NewVerifier(trustPolicy *trustpolicy.Document, pluginManager *manager.Manager) (notation.Verifier, error) {
+	if err := trustPolicy.Validate(); err != nil {
+		return nil, err
+	}
+	return &verifier{
+		trustPolicyDoc: trustPolicy,
+		pluginManager:  pluginManager,
+	}, nil
+}
+
+// Verify verifies the signature blob and returns the verified descriptor
+// upon successful verification.
+func (v *verifier) Verify(ctx context.Context, desc ocispec.Descriptor, signature []byte, opts notation.VerifyOptions) (*notation.VerificationOutcome, error) {
+	artifactRef := opts.ArtifactReference
+	envelopeMediaType := opts.SignatureMediaType
+	pluginConfig := opts.PluginConfig
+
+	trustPolicy, err := v.trustPolicyDoc.GetApplicableTrustPolicy(artifactRef)
+	if err != nil {
+		return nil, notation.ErrorNoApplicableTrustPolicy{Msg: err.Error()}
+	}
+	// ignore the error since we already validated the policy document
+	verificationLevel, _ := trustPolicy.SignatureVerification.GetVerificationLevel()
+
+	outcome := &notation.VerificationOutcome{
```

Comment:
Can we return here if verification level is skip?

Reply:
Nice catch!